First implementation of Info/Rules channel manager plugin.

GTG3000 opened the pull request:
As requested in #4 
Current functionality includes uploading, verifying and outputting multi-message files.

GTG3000 committed:
First version of the channel printer plugin.
Implemented verification and printing.

GTG3000 committed:
Implemented file uploading and document manipulation into channel printer plugin.

GTG3000 committed:
Merge branch 'channelprint'

rubellyte commented on channel_print.py:
```diff
@@ -0,0 +1,314 @@
+                wall = verify_document(self.walls[gid][wall])
+            except ValueError as e:
+                await respond(msg, f"**WARNING: {e}**")
+                return
```
This could be simplified into
```Py
except ValueError as e:
    raise CommandSyntaxError(e)
``` 
as is done in other places

rubellyte commented on channel_print.py:
```diff
@@ -0,0 +1,314 @@
+        return self
+
+
+def verify_embed(embed: dict):
```
It honestly might be best to take this out of the printer and move it to rs_utils, since it's generically useful anywhere we make embeds from user input

GTG3000 replied:
That sounds like a good idea and I've been meaning to do that for a while (except with a custom embed class).

Tested it, and I can safely make it return the EmbedDict to make it work for both JSON serialization purposes and sending via `respond(msg, embed=EmbedDict())` sort of arrangement.

Should I pack the class into the function or should I leave it separate?

rubellyte commented on channel_print.py:
```diff
@@ -0,0 +1,314 @@
+                    try:
+                        _file = urlopen(post['file'])
+                        if int(_file.info()['Content-Length']) > self.plugin_config['max_filesize']:
+                            raise ValueError("File too big.")
```
This should probably be a CommandSyntaxError or UserPermissionError

GTG3000 replied:
This is being caught by the `except` right after it. If the file is unfit, it will fail quietly and print about it via the logger.

Perhaps a better idea would be to add a file size verification before hand. For some reason, I felt like verifying file links ahead of time could not be a great idea, but I suppose it's on the users for changing the contents?

rubellyte commented on channel_print.py:
```diff
@@ -0,0 +1,314 @@
+        gid = str(msg.guild.id)
+
+        try:
+            name = msg.clean_content.split(None, 2)[1].lower()
```
Should probably be a `.split(None, 1)`

GTG3000 replied:
This is done to extract a single word from something like `"!cmd WORD all the other words we don't want here"`. Document IDs are not allowed to have spaces in them, and this ensures that.

rubellyte commented on channel_print.py:
```diff
@@ -0,0 +1,314 @@
+        gid = str(msg.guild.id)
+
+        try:
+            name = msg.clean_content.split(None, 2)[1].lower()
```
See above

rubellyte commented on channel_print.py:
```diff
@@ -0,0 +1,314 @@
+            self.walls[gid] = dict()
+
+        try:
+            name = msg.clean_content.split(None, 2)[1].lower()
```
^

rubellyte commented on channel_print.py:
````diff
@@ -0,0 +1,314 @@
+                raise CommandSyntaxError(f"Not a valid JSON file: {e}")
+        else:
+            try:
+                data = re.match(r"```.*?(?P<json>\[.+]).*?```", msg.clean_content.split(None, 2)[2], re.DOTALL)['json']
````
^

rubellyte reviewed (changes requested):
Not too much to say here but a few things could be edited

GTG3000 committed:
Implemented requested changes to channel printer.
Moved embed verification out of the plugin and into rs_utils.
Implemented a full file-check before beginning print.
